### ISSUE TITLE
VS Code extension hangs on last method

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#1819]
On this commit (https://github.com/robin-aws/libraries/commit/73cc542a337c9ac3233126c6155a1bc56be98e69) when I save the `src/Enumeration/Loopers.dfy` file and trigger verification, I can see the status bar indicator move quickly through each of the methods (love that feature btw!), but it seems to hang on the last method with the message `Verifying _default_Main...`. If I comment out that method and repeat, it hangs on the second-last method (`Example3`).

I suspect what's happening is that a "all verification complete" event is being missed somehow. If I edit the file by adding a space somewhere, for example, then the verification errors show up in the document. This isn't 100% consistent AFAICT though.

I'm using `v2.2.0` of `dafny-lang.ide-vscode` and 3f7785fb55a796e6a3e5ced5d315a0ad30001fee of Dafny `master`. I'm pulling down the latest `master` commit but doesn't look likely to fix this.

**Comment by robin-aws** [comment]
Has something to do with `SeqIteratorEnumerator`: if I comment out that class definition verification fails but at least completes without getting stuck.

**Comment by keyboardDrummer** [comment]
Could it relate to https://github.com/boogie-org/boogie/pull/497 ?

**Comment by robin-aws** [comment]
> Could it relate to https://github.com/boogie-org/boogie/pull/497 ?

I don't THINK so, because that one seems to talk about a bug where making an edit during verification causing it to get stuck. In this case I'm not doing anything concurrently, just saving the file to trigger verification and then watching the status bar.

**Comment by MikaelMayer** [comment]
I had a similar problem where I had three methods A, B and C. It was stuck verifying C which was actually trivial. When I commented out C, it was stuck verifying B the same way that you did.

I turned out that B depended on C, and when I commented C I actually had to comment where it appeared in B, so somehow it changed the verification graph.
And it's was not stuck on the verification of C, it was stuck on the verification of the termination of B, but this problem was invisible, and I found it was stuck in Z3 and I implemented a change similar to https://github.com/boogie-org/boogie/pull/497 so that at least it would get out of the loop.

But I still have the issue you mention from time to time.

**Comment by camrein** [comment]
This hanging problematic made me think of past issues. There's always this uncertainty: Is the verification really taking this long or did something crash and the status is not correctly represented in VSCode?
There are two things I'm thinking of to improve the trustworthiness of VSCode's state:

- Send a heartbeat during verifications that take exceptionally long. But I'm not sure how often it's actually the case that the state displayed frontend is getting out of sync from the language server. Therefore, this might be unnecessary complexity.
- What I believe would be interesting and beneficial in general is an error reporting mechansim: As soon as the language server encounters an error it should send a notification to VSCode to show that something went wrong. We could use the standard LSP API [ShowMessage](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#window_showMessage) to accomplish this.
